File: 0115-distinct-subsequences/0115-distinct-subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        dp = [[-1 for _ in range(len(t))] for _ in range(len(s))]
        def solve(i: int, j: int) -> int:
            if j == len(t):
                return 1
            if i == len(s):
                return 0
            if dp[i][j] == -1:
                dp[i][j] = 0
                if s[i] == t[j]:
                    dp[i][j] += solve(i + 1, j + 1)
                dp[i][j] += solve(i + 1, j)
            return dp[i][j]
        return solve(0, 0)

File: 0115-distinct-subsequences/0115-distinct-subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        if(s.length() == t.length())
        {
            return s == t;
        }
        vector<vector<int>> dp(s.length(), vector<int>(t.length(), -1));
        return solve(s, t, dp, 0, 0);
    }
private:
    int solve(string& s, string& t, vector<vector<int>>& dp, int i, int match)
    {
        if(match == t.length())
        {
            return 1;
        }
        if(i == s.length())
        {
            return 0;
        }
        if(dp[i][match] == -1)
        {
            dp[i][match] = 0;
            if(s[i] == t[match])
            {
                dp[i][match] += solve(s, t, dp, i + 1, match + 1);
            }
            dp[i][match] += solve(s, t, dp, i + 1, match);
        }
        return dp[i][match];
    }
};